benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/TestTaskScheduler.java

```diff
@@ -221,281 +245,115 @@ private synchronized void handleTestRunningState(String testID) {
-
-  }
-
-  private void determineAndExecuteExperiments(String testID) {
-
-    logger.info("determineAndExecuteExperiments with testID: " + testID);
-
-    DetermineExecuteExperimentsTask task = new DetermineExecuteExperimentsTask(testID);
-
-    // TODO - should go into a stateless queue (so that we can recover)
-    Future<?> future = taskExecutorService.submit(task);
-
-    // replace with new task
-    testTasks.put(testID, future);
-
-    // we don't wait for the task to complete since the experiment-manager
-    // will send the result
-    try {
-
-      testModelDAO.setTestRunningState(testID, HANDLE_EXPERIMENT_RESULT);
-
-    } catch (BenchFlowTestIDDoesNotExistException e) {
-      // should not happen since it was added earlier
-      logger.error("test ID does not exist - should not happen");
-    }
-  }
-
-  private void derivePredictionFunction(String testID) {
-
-    logger.info("derivePredictionFunction with testID: " + testID);
-
-    DerivePredictionFunctionTask task = new DerivePredictionFunctionTask(testID);
-
-    // TODO - should go into a stateless queue (so that we can recover)
-    Future<?> future = taskExecutorService.submit(task);
-
-    // replace with new task
-    testTasks.put(testID, future);
-
-    waitForRunningTaskToComplete(testID, future, VALIDATE_PREDICTION_FUNCTION);
-  }
-
-  private void validatePredictionFunction(String testID) {
-
-    logger.info("validatePredictionFunction with testID: " + testID);
-
-    ValidatePredictionFunctionTask task = new ValidatePredictionFunctionTask(testID);
-
-    // TODO - should go into a stateless queue (so that we can recover)
-    Future<Boolean> future = taskExecutorService.submit(task);
-
-    // replace with new task
-    testTasks.put(testID, future);
-
-    try {
-
-      boolean acceptablePredictionError = future.get();
-
-      if (acceptablePredictionError) {
-
-        testModelDAO.setTestState(testID, TERMINATED);
-        testModelDAO.setTestTerminatedState(testID, GOAL_REACHED);
-
-        // no need to execute further
-        testTasks.remove(testID);
-
-      } else {
-
-        testModelDAO.setTestRunningState(testID, REMOVE_NON_REACHABLE_EXPERIMENTS);
-
-        handleTestState(testID);
-      }
-
-    } catch (InterruptedException | ExecutionException e) {
-      // TODO - handle  properly
-      e.printStackTrace();
-    } catch (BenchFlowTestIDDoesNotExistException e) {
-      // should not happen since it was added earlier
-      logger.error("test ID does not exist - should not happen");
-    }
-  }
-
-  private void removeNonReachableExperiments(String testID) {
-
-    logger.info("removeNonReachableExperiments with testID: " + testID);
-
-    RemoveNonReachableExperimentsTask task = new RemoveNonReachableExperimentsTask(testID);
+  private void handleTerminatedState(String testID) {
+    // remove max running time timeout
+    ScheduledFuture timeoutTaskFuture = timeoutTasks.remove(testID);
+    timeoutTaskFuture.cancel(true);
```

Comment:
Not sure which case to handle. It doesn't matter if it already has been cancelled.